draft-ietf-tls-esni.md

```diff
@@ -306,6 +310,118 @@ servers to rotate the keys often and improve forward secrecy.
+extension described below. Future specifications may introduce new extensions
+and corresponding resolution algorithms.
+
+### Address Set Extension
+
+ESNIKeys records MAY point to the (set of) hosts which possess the private ESNI key
+via the following mandatory "host_pointer" ESNIKeys extension:
+  
+~~~
+    enum {
+        host_pointer(0x1002), (65535)
+    } ExtensionType;
+~~~
+
+These pointers include (1) a canonical name which, if resolved, MUST yield A
+or AAAA records that possess the corresponding private key, and (2) a set of CIDR blocks,
+each possibly of size 1, that match the IPv4 or IPv6 address of a host with the private key.
+They are encoded using the following structure.
+
+~~~~
+    enum {
+        address_v4(4),
+        address_v6(6),
+    } AddressType;
+
+    struct {
+        AddressType address_type;
+        select (address_type) {
+            case address_v4: {
+                uint8 ipv4Length;
+                opaque ipv4Address[4];
+            }
+            case address_v6: {
+                uint8 ipv6Length;
+                opaque ipv6Address[16];
+            }
+        }
+    } Address;
+
+    struct {
+        HostName name;
+        Address address_set<0..2^16-1>;
+    } HostPointer;
+~~~~
+
+name
+: The terminal name for which resolution must yield valid A or
+AAAA records pointing to a host which holds the private ESNI key.
+
+address_set
+: An optional list of Address entries, each containing a single IPv4 or IPv6
+CIDR block, possibly of length 1.
+
+Use of this structure during the ESNI resolution algorithm is described in {{esni-resolution}}.
+
+### Resolution Algorithm
+
+Clients obtain ESNI records by querying the DNS for ESNI-enabled server domains.
+In cases where the domain of the A or AAAA records being resolved do not match the
+SNI Server Name, such as when {{!RFC7838}} is being used, the alternate domain should
+be used for querying the ESNI TXT record.
+
+Clients SHOULD initiate ESNI queries in parallel alongside normal A or AAAA queries.
+The following algorithm describes a procedure by which clients can process ESNIKeys
+responses as they arrive to produce addresses for ESNI-capable hosts.
+
+~~~
+1. If an ESNIKeys response arrives before an A or AAAA response, and the ESNIKeys
+response contains a "host_pointer" extension with a full address, clients SHOULD 
+initiate TLS with ESNI to the provided address(es).
+
+2. If an A or AAAA response arrives before the ESNIKeys response, clients SHOULD
+wait up to CD seconds before initiating TLS to either address. (Clients may begin
+TCP connections in this time. QUIC connections should wait.) If an ESNIKeys
+response does not arrive in this time, clients SHOULD initiate TLS without ESNI
+to the provided address(es).
+
+3. If an ESNIKeys response with a "host_pointer" extension arrives before this time, 
+and if the address netmask matches that in the corresponding A or AAAA response, 
```

Comment:
"the address netmask" isn't defined, here or elsewhere.  I presume that this should read something like "...and if the intersection of the A or AAAA responses and the set of address ranges in address_set is not empty, then clients SHOULD initiate TLS with ESNI to addresses in that intersection."

Reply:
Oh, hah, yes!